x/liquidstake/keeper/liquidstake.go

```diff
@@ -0,0 +1,646 @@
+		Contract: k.GetParams(ctx).CwLockedPoolAddress,
+		Msg:      wasmtypes.RawContractMessage(callData),
+		Funds:    sdk.NewCoins(amount),
+	}
+
+	handler := k.router.Handler(cwMsg)
+	if handler == nil {
+		return nil, sdkerrors.ErrUnknownRequest.Wrapf("unrecognized message route: %s", sdk.MsgTypeURL(cwMsg))
+	}
+
+	msgResp, err := handler(ctx, cwMsg)
+	if err != nil {
+		return nil, types.ErrLPContract.Wrapf("error: %s, message %v", err.Error(), cwMsg)
+	}
+
+	// emit the events from the dispatched actions
+	events := msgResp.Events
+	sdkEvents := make([]sdk.Event, 0, len(events))
+	for _, event := range events {
+		e := event
+		sdkEvents = append(sdkEvents, sdk.Event(e))
+	}
+
+	ctx.EventManager().EmitEvents(sdkEvents)
+
+	return msgResp.Data, nil
+}
+
+type ExecMsg struct {
+	LockLstAssetForUser *LockLstAssetForUserMsg `json:"lock_lst_asset_for_user,omitmepty"`
+}
+
+type LockLstAssetForUserMsg struct {
+	Asset Asset  `json:"asset"`
+	User  string `json:"user"`
+}
+
+type Asset struct {
+	Amount string    `json:"amount"`
+	Info   AssetInfo `json:"info"`
+}
+
+type AssetInfo struct {
+	NativeToken NativeTokenInfo `json:"native_token"`
+}
+
+type NativeTokenInfo struct {
+	Denom string `json:"denom"`
+}
+
+// LSMDelegate captures a staked amount from existing delegation using LSM, re-stakes from proxyAcc and
+// mints stkXPRT worth of stk coin value according to NetAmount and performs LiquidDelegate.
+func (k Keeper) LSMDelegate(
+	ctx sdk.Context,
+	delegator sdk.AccAddress,
+	validator sdk.ValAddress,
+	proxyAcc sdk.AccAddress,
+	stakingCoin sdk.Coin,
+) (newShares sdk.Dec, stkXPRTMintAmount math.Int, err error) {
+	params := k.GetParams(ctx)
+
+	if params.LsmDisabled {
+		return sdk.ZeroDec(), sdk.ZeroInt(), types.ErrDisabledLSM
+	}
+
+	// check minimum liquid staking amount
+	if stakingCoin.Amount.LT(params.MinLiquidStakeAmount) {
+		return sdk.ZeroDec(), sdk.ZeroInt(), types.ErrLessThanMinLiquidStakeAmount
+	}
+
+	// check bond denomination
+	bondDenom := k.stakingKeeper.BondDenom(ctx)
+	if stakingCoin.Denom != bondDenom {
+		return sdk.ZeroDec(), sdk.ZeroInt(), sdkerrors.Wrapf(
+			types.ErrInvalidBondDenom, "invalid coin denomination: got %s, expected %s", stakingCoin.Denom, bondDenom,
+		)
+	}
+
+	whitelistedValsMap := types.GetWhitelistedValsMap(params.WhitelistedValidators)
+	activeVals := k.GetActiveLiquidValidators(ctx, whitelistedValsMap)
+	if activeVals.Len() == 0 || !activeVals.TotalWeight(whitelistedValsMap).IsPositive() {
+		return sdk.ZeroDec(), sdk.ZeroInt(), types.ErrActiveLiquidValidatorsNotExists
+	}
+
+	if !whitelistedValsMap.IsListed(validator.String()) {
+		return sdk.ZeroDec(), sdk.ZeroInt(), types.ErrLiquidValidatorsNotExists.Wrap("delegation from a non allowed validator")
+	}
+
+	// NetAmount must be calculated before send
+	nas := k.GetNetAmountState(ctx)
+
+	// perform an LSM tokenize->bank send->redeem flow: moving delegation from user's account onto proxyAcc
+
+	lsmTokenizeMsg := &stakingtypes.MsgTokenizeShares{
```

Comment:
I feel the module should receive tokenized shares already (or accept both). Although majority of users will do everything from our UI, it gives more flexibility to the module. The tokenizing process can all be done from our UI. Now this will mean 2 txs but in the future with hooks it can be done with just one.

Reply:
> I feel the module should receive tokenized shares already

We abstract away from tokenized shares of LSM, but it's doable, I am not sure who will have them though. And why xD

Reply:
> The tokenizing process 

Tokenizing there is hidden within the module just to move stake. And we don't separate these because stake must be destined to LP.